File: backend/app/domain/models/canvas.py
"""Canvas domain models for the interactive AI canvas editor."""

from datetime import UTC, datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, field_validator


class ElementType(str, Enum):
    RECTANGLE = "rectangle"
    ELLIPSE = "ellipse"
    TEXT = "text"
    IMAGE = "image"
    LINE = "line"
    PATH = "path"
    GROUP = "group"


class FillType(str, Enum):
    SOLID = "solid"
    GRADIENT = "gradient"


class GradientType(str, Enum):
    LINEAR = "linear"
    RADIAL = "radial"


class ExportFormat(str, Enum):
    PNG = "png"
    SVG = "svg"
    JSON = "json"


class SolidFill(BaseModel):
    type: FillType = FillType.SOLID
    color: str = "#000000"


class GradientStop(BaseModel):
    offset: float = 0.0  # 0-1
    color: str = "#000000"


class GradientFill(BaseModel):
    type: FillType = FillType.GRADIENT
    gradient_type: GradientType = GradientType.LINEAR
    stops: list[GradientStop] = Field(default_factory=list)
    angle: float = 0.0  # degrees, for linear gradients


Fill = SolidFill | GradientFill


class Stroke(BaseModel):
    color: str = "#000000"
    width: float = 1.0
    dash: list[float] | None = None  # dash pattern e.g. [5, 5]


class Shadow(BaseModel):
    color: str = "rgba(0,0,0,0.3)"
    blur: float = 10.0
    offset_x: float = 5.0
    offset_y: float = 5.0


class TextStyle(BaseModel):
    font_family: str = "Arial"
    font_size: float = 16.0
    font_weight: str = "normal"  # "normal", "bold", etc.
    font_style: str = "normal"  # "normal", "italic"
    text_align: str = "left"  # "left", "center", "right"
    vertical_align: str = "top"  # "top", "middle", "bottom"
    line_height: float = 1.2
    letter_spacing: float = 0.0
    text_decoration: str = "none"  # "none", "underline", "line-through"


class CanvasElement(BaseModel):
    """Single canvas element with flattened type-specific fields."""

    id: str
    type: ElementType
    name: str = ""
    x: float = 0.0
    y: float = 0.0
    width: float = 100.0
    height: float = 100.0
    rotation: float = 0.0
    scale_x: float = 1.0
    scale_y: float = 1.0
    opacity: float = 1.0
    visible: bool = True
    locked: bool = False
    z_index: int = 0
    fill: dict[str, Any] | None = None  # SolidFill or GradientFill as dict
    stroke: dict[str, Any] | None = None  # Stroke as dict
    shadow: dict[str, Any] | None = None  # Shadow as dict
    corner_radius: float = 0.0

    # Text-specific
    text: str | None = None
    text_style: dict[str, Any] | None = None  # TextStyle as dict

    # Image-specific
    src: str | None = None

    # Path/Line-specific
    points: list[float] | None = None  # [x1, y1, x2, y2, ...]

    # Group-specific
    children: list[str] | None = None  # child element IDs


class CanvasPage(BaseModel):
    """A single page/artboard in the canvas project."""

    id: str
    name: str = "Page 1"
    width: float = 1920.0
    height: float = 1080.0
    background: str = "#FFFFFF"
    elements: list[CanvasElement] = Field(default_factory=list)
    sort_order: int = 0


class CanvasProject(BaseModel):
    """A canvas project owned by a user."""

    id: str
    user_id: str
    session_id: str | None = None  # links to agent session if created by agent
    name: str = "Untitled Design"
    description: str = ""
    pages: list[CanvasPage] = Field(default_factory=list)
    width: float = 1920.0
    height: float = 1080.0
    background: str = "#FFFFFF"
    thumbnail: str | None = None  # base64 or URL
    version: int = 1
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        v = v.strip()
        if len(v) < 1 or len(v) > 200:
            raise ValueError("Name must be between 1 and 200 characters")
        return v

    @field_validator("pages")
    @classmethod
    def validate_pages(cls, v: list[CanvasPage]) -> list[CanvasPage]:
        if len(v) > 50:
            raise ValueError("Maximum 50 pages per project")
        return v


class CanvasVersion(BaseModel):
    """A saved version/snapshot of a canvas project."""

    id: str
    project_id: str
    version: int
    name: str = ""
    pages: list[CanvasPage] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))

File: frontend/src/types/canvas.ts
/**
 * Canvas editor TypeScript types.
 * Mirrors backend domain models for canvas projects.
 */

export type ElementType = 'rectangle' | 'ellipse' | 'text' | 'image' | 'line' | 'path' | 'group'

export type EditorTool =
  | 'select'
  | 'rectangle'
  | 'ellipse'
  | 'text'
  | 'image'
  | 'pen'
  | 'line'
  | 'hand'

export type FillType = 'solid' | 'gradient'
export type GradientType = 'linear' | 'radial'

export interface SolidFill {
  type: 'solid'
  color: string
}

export interface GradientStop {
  offset: number
  color: string
}

export interface GradientFill {
  type: 'gradient'
  gradient_type: GradientType
  stops: GradientStop[]
  angle: number
}

export type Fill = SolidFill | GradientFill

export interface Stroke {
  color: string
  width: number
  dash?: number[]
}

export interface Shadow {
  color: string
  blur: number
  offset_x: number
  offset_y: number
}

export interface TextStyle {
  font_family: string
  font_size: number
  font_weight: string
  font_style: string
  text_align: string
  vertical_align: string
  line_height: number
  letter_spacing: number
  text_decoration: string
}

export interface CanvasElement {
  id: string
  type: ElementType
  name: string
  x: number
  y: number
  width: number
  height: number
  rotation: number
  scale_x: number
  scale_y: number
  opacity: number
  visible: boolean
  locked: boolean
  z_index: number
  fill?: Fill
  stroke?: Stroke
  shadow?: Shadow
  corner_radius: number
  text?: string
  text_style?: TextStyle
  src?: string
  points?: number[]
  children?: string[]
}

export interface CanvasPage {
  id: string
  name: string
  width: number
  height: number
  background: string
  elements: CanvasElement[]
  sort_order: number
}

export interface CanvasProject {
  id: string
  user_id: string
  session_id: string | null
  name: string
  description: string
  pages: CanvasPage[]
  width: number
  height: number
  background: string
  thumbnail: string | null
  version: number
  created_at: string
  updated_at: string
}

export interface CanvasProjectSyncState {
  session_id: string | null
  version: number
  updated_at: string
}

export type CanvasSyncStatus = 'live' | 'saved' | 'syncing' | 'stale' | 'conflict'

export interface CanvasRemoteSyncState {
  sessionId: string | null
  serverVersion: number
  pendingRemoteVersion: number | null
  hasRemoteConflict: boolean
  isStale: boolean
  lastRemoteOperation: string | null
  lastRemoteSource: 'agent' | 'manual' | 'system' | null
  lastChangedElementIds: string[]
  highlightedElementIds: string[]
}

export interface CanvasVersion {
  id: string
  project_id: string
  version: number
  name: string
  created_at: string
}

export interface EditorState {
  activeTool: EditorTool
  activePageIndex: number
  selectedElementIds: string[]
  zoom: number
  panX: number
  panY: number
  showGrid: boolean
  snapEnabled: boolean
  isDirty: boolean
}
